Automattic/jetpack
# Add/device detection js package

Add stripped down version of the [device-detector-js](https://www.npmjs.com/package/device-detector-js) npm package. This is to be moved to `js-packages` later on

( Don't be alarmed by the 40000 lines of added code, the vast majority of that is JSON files for matching user agents )

NOTE:
I stripped this down as much as possible, it seems as though browser type and version are used in certain circumstances to detect the type and brand of the device so that had to stay in (specifically the code was having trouble detecting tablets when I took browser parsing out)

#### Changes proposed in this Pull Request:
Add an easy to use package to detect device type and brand within the plugin. The use case for this initial import is to display a QR code when users are using a desktop, an Apple Store link when using an Apple product, and a Google Play store link if using a different brand smartphone/tablet

#### Other information:


#### Jetpack product discussion
P2: pbNhbs-3od-p2
Slack chat: p1661896651864719-slack-CBG1CP4EN

#### Does this pull request change what data or activity we track or use?
This detects device type, but does not track it, so no it does now

#### Testing instructions:
1. Checkout this branch via `git switch add/device-detection-js-package`

2. Head to the `projects/plugins/jetpack/_inc/client/recommendations/index.jsx` file and add the following import to the imports list
`import DeviceDetector from './packages/device-detector-js';`
and then add the following lines of code to the beginning of the `RecommendationsComponent` function
```
const deviceDetector = new DeviceDetector();
const userAgent = window.navigator.userAgent;
const device = deviceDetector.parse( userAgent );

console.log( device );
```

![Screen Shot 2022-09-01 at 11 56 34 AM](https://user-images.githubusercontent.com/65001528/187981520-d8c2f1a8-1c2a-4ffa-9a51-62e4e394ceb5.jpg)

3. Spin up your local environment and head to `/wp-admin/admin.php?page=jetpack#/recommendations/summary`

4. Open up the dev tools and go to the console. You should see the device you are using printed to the console. Make sure it is accurate

![image](https://user-images.githubusercontent.com/65001528/187981769-0e052243-f737-48a0-8ebe-c54c748f3f3e.png)

5. To test additional user agents, you can ( in the Chrome Dev tools, not sure how to do this on other browsers ) select the vertical ellipsis > More Tools > Network Conditions

![image](https://user-images.githubusercontent.com/65001528/187984843-87091bc6-6501-44d0-b4f6-3977ffee798c.png)

6. Here, you can select different user agents from a dropdown by unselecting `Use browser default` and clicking on the dropdown below

![image](https://user-images.githubusercontent.com/65001528/187982148-71b64456-f0ca-42c6-82d7-c4516398380a.png)

7. Try out several of these to make sure the device returned is still accurate. I'll post screenshots of an example below

![image](https://user-images.githubusercontent.com/65001528/187982323-ed790fd5-dec9-46ed-9ba4-ca669f2348aa.png)

![image](https://user-images.githubusercontent.com/65001528/187982380-5b755621-778a-4191-afbc-a8f76b44ef2d.png)

8. If you want more devices to test on, you can use BrowserStack. Automattic has accounts to use so you don't have to pay for it. More info here: PCYsg-iPs-p2 and here: 2d7c8-pb

I tested this out on apple and android tablets, apple and android phones, and windows browsers and it seemed to do well

If there is any other way you can think of to stress test this, feel free to do so!
